virtcontainers/container.go

```diff
@@ -1267,6 +1267,56 @@ func (c *Container) updateResources(oldResources, newResources ContainerResource
 		// recalculate the actual number of vCPUs if a different number of vCPUs was removed
 		newResources.VCPUs = oldVCPUs - vcpusRemoved
 	}
+	return nil
+}
+
+func (c *Container) updateMemoryResources(oldResources, newResources ContainerResources) error {
+	oldMem := oldResources.Mem
+	newMem := newResources.Mem
+
+	if newMem == 0 || oldMem == newMem {
+		c.Logger().WithFields(logrus.Fields{
+			"old-mem": fmt.Sprintf("%dMB", oldMem),
+			"new-mem": fmt.Sprintf("%dMB", newMem),
+		}).Debug("the actual number of Mem will not be modified")
+		return nil
+	}
+
+	if oldMem < newMem {
+		// hot add memory
+		addMemMB := newMem - oldMem
+		// hot add memory must be aligned to 128MB
```

Comment:
this is a limitation in the kernel https://elixir.bootlin.com/linux/v4.18.4/source/include/linux/mmzone.h#L1082 (PAGE_SECTION_MASK), we can deal with this limitation but we need support for memory align in QEMU, I think those patches haven't been merged yet (not sure).

Reply:
Yes, I don't find any limitation of memory section align in QEMU, so I do judgment here. I will define a variable of 128 and create a function to encapsulate the memory checking, thanks.

Reply:
@devimc @jodh-intel The reason why hot add memory must be aligned to 128MB is that guestos will check whether hotplug memory range is aligned with memory section when we hot add memory to the guest, with reference to https://elixir.bootlin.com/linux/v4.18.4/source/mm/memory_hotplug.c#L1075.
On x86_64 the size of section is 128MB, but on other platform section may not be 128MB. If we want to keep compatible with other platform, we need to read the section size of guest rather than use 128MB.

Reply:
@clarecch @jodh-intel : You are right. For ppc64le, the hot plugged memory must be aligned to 256MB.

Reply:
@nitkon @devimc @jodh-intel We have some discuss about hot add memory align in https://github.com/kata-containers/runtime/pull/580

Reply:
@clarecch - rather than https://github.com/kata-containers/agent/issues/354, is there are reason that we can't just update the agent to align to the correct boundary when it's given a memory adjustment by the runtime? That way we don't need a new gRPC call and all the logic is encapsulated in the agent.

Reply:
@jodh-intel The qmp command that hotplug memory needs a memory size aligned to correct boundary, so runtime should know the memory block, which gets from the agent.